client/src/main/java/org/apache/celeborn/client/read/WorkerPartitionReader.java

```diff
@@ -74,25 +82,54 @@ public void onSuccess(int chunkIndex, ManagedBuffer buffer) {
           }
 
           @Override
-          public void onFailure(int chunkIndex, Throwable e) {
-            String errorMsg = "Fetch chunk " + chunkIndex + " failed.";
-            logger.error(errorMsg, e);
-            exception.set(new IOException(errorMsg, e));
+          public void onFailure(int chunkIndex, Throwable e, PartitionLocation location) {
+            readableLocations.remove(location);
+            if (readableLocations.isEmpty()) {
+              String errorMsg = "Fetch chunk " + chunkIndex + " failed.";
+              logger.error(errorMsg, e);
+              exception.set(new IOException(errorMsg, e));
+            } else {
+              if (!readableLocations.contains(location)) {
```

Comment:
Seems this condition will always be true because readableLocations removes location first